bitswap/network/httpnet/msg_sender.go

```diff
@@ -0,0 +1,575 @@
+// which is forwarded to the receivers (in a separate goroutine).
+func (sender *httpMsgSender) SendMsg(ctx context.Context, msg bsmsg.BitSwapMessage) error {
+	// SendMsg gets called from MessageQueue and returning an error
+	// results in a MessageQueue shutdown. Errors are only returned when
+	// we are unable to obtain a single valid Block/Has response. When a
+	// URL errors in a bad way (connection, 500s), we continue checking
+	// with the next available one.
+
+	// unless we have a wantlist, we bailout.
+	wantlist := msg.Wantlist()
+	if len(wantlist) == 0 {
+		return nil
+	}
+
+	// Keep metrics of wantlists sent and how long it took
+	sender.ht.metrics.WantlistsTotal.Inc()
+	sender.ht.metrics.WantlistsItemsTotal.Add(float64(len(wantlist)))
+	log.Debugf("sending wantlist: %s (%d items)", sender.peer, len(wantlist))
+	now := time.Now()
+	defer func() {
+		sender.ht.metrics.WantlistsSeconds.Observe(float64(time.Since(now)) / float64(time.Second))
+	}()
+
+	// This is mostly a cosmetic action since the bitswap.Client is just
+	// logging the errors.
+	sendErrors := func(err error) {
+		if err != nil {
+			for _, recv := range sender.ht.receivers {
+				recv.ReceiveError(err)
+			}
+		}
+	}
+
+	var err error
+
+	// obtain contexts for all the entries in the wantlits.  This allows
+	// us to react when cancels arrive to wantlists that we are going
+	// through.  We use a Background context because requests will be
+	// ongoing when we return and the parent context is cancelled.
+	parentCtx := context.Background()
+	entryCtxs := make([]context.Context, len(wantlist))
+	entryCancels := make([]context.CancelFunc, len(wantlist))
+	nop := func() {}
+	for i, entry := range wantlist {
+		if entry.Cancel {
+			entryCtxs[i] = ctx
+			entryCancels[i] = nop
+		} else {
+			entryCtxs[i], entryCancels[i] = sender.ht.requestTracker.requestContext(parentCtx, entry.Cid)
+		}
+	}
+
+	var wgRequests sync.WaitGroup
+	resultsCollector := make(chan httpResult, len(wantlist))
+
+WANTLIST_LOOP:
+	for i, entry := range wantlist {
+		reqStart := time.Now()
+
+		if entry.Cancel { // shortcut cancel entries.
+			sender.ht.requestTracker.cancelRequest(entry.Cid)
+			sender.ht.metrics.updateStatusCounter(0)
+			// Do not observe request time for cancel requests as
+			// they cost us nothing, so it is unfair to compare
+			// against bsnet requests-time.
+			// sender.ht.metrics.RequestTime.Observe(float64(time.Since(reqStart))
+			// / float64(time.Second))
+			continue
+		}
+
+		resultCh := make(chan httpResult, 1)
+
+		reqInfo := httpRequestInfo{
+			ctx:    entryCtxs[i],
+			sender: sender,
+			entry:  entry,
+			result: resultCh,
+		}
+
+		select {
+		case <-ctx.Done():
+			// our context cancelled so we must abort.
+			err = ctx.Err()
+			break WANTLIST_LOOP
+		case sender.ht.httpRequests <- reqInfo:
+			wgRequests.Add(1)
+			go func(start time.Time) {
+				defer wgRequests.Done()
+				result := <-resultCh
+				close(resultCh)
+				resultsCollector <- result
```

Comment:
Does not look like `close(resultCh)` is needed.
```suggestion
				resultsCollector <-resultCh
```

Reply:
When I was a kid I was taught to close my channels. So I'm guessing now they will be properly cleaned up even if left open?

Reply:
When a channel is closed, that is an indication that there is a reader that might otherwise not know to stop reading. Closing is just a signal to readers and does not trigger any resource cleanup or earlier GC. They get cleaned up the same.

If you prefer to closing the channel anyway as future-proofing, that's fine.